web/app/controllers/Application.java

```diff
@@ -170,32 +175,48 @@ public static Result login()
         return ok(login.render(csrfToken, isInternal, piwikSiteId));
     }
 
-    public static Result authenticate()
-    {
-        DynamicForm loginForm = form().bindFromRequest();
-        String username = loginForm.get("username");
-        String password = loginForm.get("password");
+    @BodyParser.Of(BodyParser.Json.class)
+    public static Result authenticate() {
+        // Create a new reponse ObjectNode to return when authenticate
+        //   requst is successful
+        ObjectNode response = Json.newObject();
+        JsonNode json = request().body().asJson();
+        // Extract username and password as String from JsonNode,
+        //   null if they are not strings
+        String username = json.findPath("username").textValue();
+        String password = json.findPath("password").textValue();
 
-        if (StringUtils.isBlank(username) || StringUtils.isBlank(password))
-        {
-            flash("error", "Invalid username or password");
-            return redirect(controllers.routes.Application.login());
+        if (username == null || StringUtils.isBlank(username)) {
+            return badRequest("Missing or invalid [username]");
+        }
+        if (password == null || StringUtils.isBlank(password)) {
+            return badRequest("Missing or invalid [credentials]");
         }
 
-        try
-        {
-            AuthenticationManager.authenticateUser(username, password);
+        // Create a uuid string for this session if one doesn't already exist
+        //   to be appended to the Result object
+        String uuid = session("uuid");
+        if (uuid == null) {
+            uuid = java.util.UUID.randomUUID().toString();
+            session("uuid", uuid);
         }
-        catch (Exception e)
-        {
-            Logger.error("Authentication failed for user " + username);
-            Logger.error(e.getMessage());
-            flash("error", "Invalid username or password");
-            return redirect(controllers.routes.Application.login());
+
+        try {
+            AuthenticationManager.authenticateUser(username, password);
+        } catch (Exception e) {
+            return badRequest("Invalid credentials");
         }
+
         session().clear();
```

Comment:
Maybe move the "set session actions" to behind this clear(). Also need to set session "user".